### PR TITLE
Feature/update libs

### PR DESCRIPTION
**Pull request by samleeflang**
Update all libraries for ELViS to the latest version.
Had to rewrite the authentication and authorisation code.
Fixed a bunch of Sonar issues.

### CODE REVIEW

File: src/auth/ktor/Principal.kt [hunk 1]
```diff
@@ -16,12 +16,12 @@ import org.synthesis.keycloak.admin.KeycloakUserGroup
  */
 internal fun JWTCredential.principal(realm: KeycloakRealm, client: KeycloakApiClient): UserAccount = UserAccount(
     id = id(),
-    email = payload.getClaim("email").asString(),
+    email = "",
     groups = groups(),
```
Comment (southeo) [inline src/auth/ktor/Principal.kt:20]: Is this left blank because we might not get the email from keycloak/the sign in partner anymore?

**Review by southeo — APPROVED** [review]

File: src/auth/ktor/Principal.kt (hunk 1 above)
Comment (samleeflang) [reply src/auth/ktor/Principal.kt:20]: Yes that is correct. Part of the information this expects is no longer in the token. Which is not an issue as it doesn't actually do anything with it. It only retrieves the id from this and then gathers all user information from the database (see SecurityConfiguration). I create a task on the backlog to change this implementation https://naturalis.atlassian.net/browse/DD-152

File: src/WebServer.kt [hunk 2]
```diff
@@ -1,18 +1,8 @@
...
 
-import io.ktor.locations.KtorExperimentalLocationsAPI
-import io.ktor.server.engine.embeddedServer
-import io.ktor.server.netty.Netty
-import io.ktor.util.InternalAPI
-import io.ktor.util.KtorExperimentalAPI
-import kotlinx.coroutines.ExperimentalCoroutinesApi
-import kotlinx.coroutines.ObsoleteCoroutinesApi
+import io.ktor.server.engine.*
+import io.ktor.server.netty.*
```
Comment (TomDijkema) [inline src/WebServer.kt:4]: Were most these libraries deprecated (as they are removed in all files)? Or were they unnecessary for production?

File: src/account/Routes.kt [hunk 3]
```diff
@@ -36,8 +37,13 @@ fun Route.profileRoutes() {
...
                     attributes = currentUser.attributes.copy(
-                        gender = request.gender?.let { Gender.valueOf(it.toUpperCase()) } ?: Gender.OTHER,
-                        relatedInstitutionId = request.relatedInstitutionId?.let { InstitutionId.fromString(it) },
+                        gender = request.gender?.let { Gender.valueOf(it.uppercase()) }
+                            ?: Gender.OTHER,
+                        relatedInstitutionId = request.relatedInstitutionId?.let {
+                            InstitutionId.fromString(
+                                it
+                            )
+                        },
```
Comment (TomDijkema) [inline src/account/Routes.kt:46]: Much better to read, they loved to make long one liners

File: src/auth/ktor/RoleBasedAuthPlugin.kt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+
+class RoleBasedAuthPluginConfiguration {
+    var roleExtractor: ((Principal) -> Set<Role>)? = null
+        private set
+
+    fun extractRoles(extractor: (Principal) -> Set<Role>) {
+        roleExtractor = extractor
+    }
+}
```
Comment (TomDijkema) [inline src/auth/ktor/RoleBasedAuthPlugin.kt:117]: It seems like you extracted the authentication roles part from different files and compiled them to one single file to make it more clear?

File: src/environment/DefaultUserImporter.kt [hunk 5]
```diff
@@ -22,7 +23,7 @@ class DefaultUserImporter(
             attributes = UserAccountAttributes(
                 orcId = OrcId("0000-0000-0000-0000"),
                 institutionId = null,
-                relatedInstitutionId = InstitutionId.fromString("grid.35937.3b"),
+                relatedInstitutionId = institutionId,
```
Comment (TomDijkema) [inline src/environment/DefaultUserImporter.kt:26]: What is the purpose of the default user and use of the static identifier?

File: src/WebServer.kt (hunk 2 above)
Comment (samleeflang) [reply src/WebServer.kt:4]: No most of the libraries moved from ktor.X to ktor.server which impacted a lot of files. So most depencies are the same but they have just been renamed in a newer version

File: src/auth/ktor/RoleBasedAuthPlugin.kt (hunk 4 above)
Comment (samleeflang) [reply src/auth/ktor/RoleBasedAuthPlugin.kt:117]: Not really. Before they used a deprecated plugin for ktor. This is the new way of adding pipeline plugins. I also added functionality for withRoles or withoutRoles which are not necessarily used but are possible this way.

File: src/environment/DefaultUserImporter.kt (hunk 5 above)
Comment (samleeflang) [reply src/environment/DefaultUserImporter.kt:26]: No idea but I didn't dare to remove it. I did make the string to a static variable as this was a Sonar issue and easy to fix

**Review by TomDijkema — APPROVED** [review]

**Pull request merged**